### PR TITLE
Add support for version 2 XML key files

### PR DESCRIPTION
**Pull request by phoerious**
As discussed in #4317, the next KeePass2 release will ship with support for a new generation of XML key files which enable hash integrity checks.

This patch adds support for reading and generating this new format. By default, KeePass2 now uses the .keyx extension for generated key files, which was added to KeePassXC's key generation file chooser filter. We continue to generate hashed binary key files by default, but the user can explicitly save the file with the new .keyx extension to generate an XML v2 key file (currently undocumented).

When opening a database, the key file type is still determined by content negotation, so the file extension has no impact here.

As an additional change, the legacy key file warnings have been improved slightly to be less confusing and more helpful.

**This change should be backported to 2.6.3.**

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ New feature (change that adds functionality)

### CODE REVIEW

**Comment by phoerious** [comment]
@DReichl KeePass2 seems to always add XML schema attributes to its key files, but can read them just fine when they are missing. Is there any specific reason for that? There is no linked schema anyway, so those attributes are nothing but wasted bytes.

File: src/gui/DatabaseOpenWidget.cpp [hunk 1]
```diff
@@ -301,12 +301,14 @@ QSharedPointer<CompositeKey> DatabaseOpenWidget::buildDatabaseKey()
...
         }
-        if (key->type() != FileKey::Hashed && !config()->get(Config::Messages_NoLegacyKeyFileWarning).toBool()) {
+        if (key->type() != FileKey::KeePass2XMLv2 && key->type() != FileKey::Hashed
+            && !config()->get(Config::Messages_NoLegacyKeyFileWarning).toBool()) {
             QMessageBox legacyWarning;
             legacyWarning.setWindowTitle(tr("Legacy key file format"));
-            legacyWarning.setText(tr("You are using a legacy key file format which may become\n"
-                                     "unsupported in the future.\n\n"
-                                     "Please consider generating a new key file."));
+            legacyWarning.setText(tr("You are using a legacy key file format which may become<br>"
```
Comment (droidmonkey) [inline src/gui/DatabaseOpenWidget.cpp:308]: Given Dominik's feedback, recommend replacing the term "legacy" with "older" or similar.

Personally I think we should remove this message on the database unlock screen and just show it when adding a key to the credentials. At this point everyone who has a legacy key has seen the message already and either doesn't care, ignored it, or changed their key.

File: src/gui/databasekey/KeyFileEditWidget.cpp [hunk 2]
```diff
@@ -47,12 +47,12 @@ bool KeyFileEditWidget::addToCompositeKey(QSharedPointer<CompositeKey> key)
         return false;
     }
 
-    if (fileKey->type() != FileKey::Hashed) {
+    if (fileKey->type() != FileKey::KeePass2XMLv2 && fileKey->type() != FileKey::Hashed) {
         QMessageBox::warning(getMainWindow(),
                              tr("Legacy key file format"),
-                             tr("You are using a legacy key file format which may become\n"
+                             tr("You selected a legacy key file format which may become\n"
```
Comment (droidmonkey) [inline src/gui/databasekey/KeyFileEditWidget.cpp:53]: Same as above.

File: src/keys/FileKey.cpp [hunk 3]
```diff
@@ -218,87 +274,81 @@ bool FileKey::create(const QString& fileName, QString* errorMsg, int size)
...
+                }
+            }
         } else if (xmlReader.name() == "Key") {
-            data = loadXmlKey(xmlReader);
+            while (!xmlReader.error() && xmlReader.readNextStartElement()) {
+                if (xmlReader.name() == "Data") {
+                    keyFileData.hash = QByteArray::fromHex(xmlReader.attributes().value("Hash").toLatin1());
+                    QByteArray rawData = xmlReader.readElementText().simplified().replace(" ", "").toLatin1();
+
+                    if (keyFileData.version == "1.00" && Tools::isBase64(rawData)) {
```
Comment (droidmonkey) [inline src/keys/FileKey.cpp:319]: We should add error message for the "else" case, neither the version matches or the raw data is not base64/hex.

**Review by droidmonkey — CHANGES_REQUESTED** [review]

File: src/keys/FileKey.cpp (hunk 3 above)
Comment (phoerious) [reply src/keys/FileKey.cpp:319]: In that case it would fall through and test the next format.

Comment (phoerious) [reply src/keys/FileKey.cpp]: I decided to go for an error message. May fail in the rare case that somebody was using an invalid key file that looks like an XML key file, but isn't, but overall it's cleaner.

**Review by droidmonkey — APPROVED** [review]

**Pull request merged**